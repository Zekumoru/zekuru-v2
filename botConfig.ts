export default {
  messageCharactersLimit: 2000,
};
