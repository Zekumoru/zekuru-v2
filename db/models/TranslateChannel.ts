import * as deepl from 'deepl-node';
import { Schema, Types, model } from 'mongoose';

interface ITranslateChannelSchema {
  id: string;
  guildId: string;
  sourceLang: deepl.SourceLanguageCode;
  targetLang: deepl.TargetLanguageCode;
  createdAt: Date;
}

export interface ITranslateChannel extends ITranslateChannelSchema {
  _id: Types.ObjectId;
}

const TranslateChannelSchema = new Schema<ITranslateChannelSchema>({
  id: {
    type: String,
    required: true,
  },
  guildId: {
    type: String,
    required: true,
  },
  sourceLang: {
    type: String,
    required: true,
  },
  targetLang: {
    type: String,
    required: true,
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

export default model('TranslateChannel', TranslateChannelSchema);
