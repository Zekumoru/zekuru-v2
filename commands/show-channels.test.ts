import showChannels from './show-channels';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';
import TranslateChannel from '../db/models/TranslateChannel';
import {
  clearSampleChannelLinks,
  resetSampleChannelLinks,
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';
import botConfig from '../botConfig';

jest.mock('../cache');
jest.mock('../db/models/TranslateChannel.ts');
jest.mock(
  '../botConfig.ts',
  () => ({ messageCharactersLimit: 500 } as typeof botConfig)
);

describe('/show-channels command', () => {
  const mockTranslateChannel = jest.mocked(TranslateChannel);

  afterEach(() => {
    clearSampleChannelLinks();
    jest.clearAllMocks();
  });

  it(`should notify user that this command is only on servers`, async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await showChannels.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user when they are no translate channels to show`, async () => {
    const interaction = createMockInteraction();
    mockTranslateChannel.find.mockResolvedValue([]);

    await showChannels.execute(interaction);

    expect(mockTranslateChannel.find).toHaveBeenCalledWith({
      guildId: interaction.guildId,
    });
    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should show channels`, async () => {
    const interaction = createMockInteraction();
    mockTranslateChannel.find.mockResolvedValue(sampleTranslateChannels);
    sampleChannelLinks[0].links.push(sampleTranslateChannels[1] as any);
    sampleChannelLinks[0].links.push(sampleTranslateChannels[2] as any);
    sampleChannelLinks[1].links.push(sampleTranslateChannels[0] as any);
    sampleChannelLinks[1].links.push(sampleTranslateChannels[2] as any);
    sampleChannelLinks[2].links.push(sampleTranslateChannels[0] as any);
    sampleChannelLinks[2].links.push(sampleTranslateChannels[1] as any);

    await showChannels.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should show embeds instead if the message is over Discord's character limit`, async () => {
    const interaction = createMockInteraction();
    mockTranslateChannel.find.mockResolvedValue(sampleTranslateChannels);
    const guildId = sampleChannelLinks[0].guildId;
    sampleTranslateChannels.push({
      guildId,
      id: '000',
      sourceLang: 'it',
      targetLang: 'it',
    });
    sampleTranslateChannels.push({
      guildId,
      id: '001',
      sourceLang: 'fr',
      targetLang: 'fr',
    });
    sampleTranslateChannels.push({
      guildId,
      id: '002',
      sourceLang: 'es',
      targetLang: 'es',
    });

    await showChannels.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    resetSampleChannelLinks();
  });
});
