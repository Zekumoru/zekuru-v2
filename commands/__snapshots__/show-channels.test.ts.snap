// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/show-channels command should notify user that this command is only on servers 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`/show-channels command should notify user when they are no translate channels to show 1`] = `
{
  "content": "There are no translate channels yet. You can start setting them by using the \`/set\` command.",
}
`;

exports[`/show-channels command should show channels 1`] = `
{
  "content": "**Showing all translate channels**
There are a total of 4 translate channels.
- <#100> is set to \`en\` and is linked to 2 translate channels.
- <#200> is set to \`ja\` and is linked to 2 translate channels.
- <#300> is set to \`zh\` and is linked to 2 translate channels.
- <#400> is set to \`ko\` and is **not linked** to any translate channels.",
}
`;

exports[`/show-channels command should show embeds instead if the message is over Discord's character limit 1`] = `
{
  "embeds": [
    {
      "color": 39423,
      "description": "**Showing all translate channels**
There are a total of 7 translate channels.
- <#100> is set to \`en\` and is **not linked** to any translate channels.
- <#200> is set to \`ja\` and is **not linked** to any translate channels.
- <#300> is set to \`zh\` and is **not linked** to any translate channels.
- <#400> is set to \`ko\` and is **not linked** to any translate channels.
- <#000> is set to \`it\` and is **not linked** to any translate channels.
- <#001> is set to \`fr\` and is **not linked** to any translate channels.
- <#002> is set to \`es\` and is **not linked** to any translate channels.",
    },
  ],
}
`;
